cwlVersion: v1.0
class: CommandLineTool
label: "eXpress: Streaming quantification for high-throughput sequencing"
doc: "Streaming quantification for high-throughput sequencing. https://pachterlab.github.io/eXpress/manual.html"

hints:
  DockerRequirement:
    dockerPull: yyabuki/express:1.5.1

baseCommand: [express]

arguments:
  - prefix: -o
    valueFrom: $(runtime.outdir)

inputs:
  target_fasta:
    label: "A multi-FASTA file of target sequences for which the abundances will be measured"
    doc: "A multi-FASTA file of target sequences for which the abundances will be measured. In the case of RNA-Seq, these are the transcript sequences. If the transcriptome of your organism is not annotated, you can generate this file from your sequencing reads using a de novo transcriptome assembler such as Trinity, Oases, or Trans-ABySS. If your organism has a reference genome, you can assemble transcripts directly from mapped reads using Cufflinks. If your genome is already annotated (in GTF/GFF), you can generate a multi-FASTA file using the UCSC Genome Browser by uploading your annotation as a track and downloading the sequences under the 'Tables' tab."
    type: File
    inputBinding:
      position: 1
  input_bam:
    label: "A SAM or binary SAM (BAM) alignments"
    doc: "A file, multiple files, or a piped stream of SAM or binary SAM (BAM) alignments as input. The SAM alignments should be generated by mapping your sequencing reads to the target sequences specified in the multi-FASTA input file described above."
    type: File
    inputBinding:
      position: 2

outputs:
  express_result:
    type:
      type: array
      items: File
    outputBinding:
      glob: "*"

$namespaces:
  s: https://schema.org/
  edam: http://edamontology.org/

s:license: https://spdx.org/licenses/Apache-2.0
s:codeRepository: https://github.com/pitagora-network/pitagora-cwl

$schemas:
  - https://schema.org/docs/schema_org_rdfa.html
  - http://edamontology.org/EDAM_1.18.owl
